drivers/rf-transceiver/hmc630x/iio_hmc630x.c

```diff
@@ -0,0 +1,378 @@
+		if (ret)
+			return ret;
+		break;
+	default:
+		return -EINVAL;
+	};
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, (int32_t *)&temp);
+}
+
+static struct iio_attribute hmc630x_iio_temp_attrs[] = {
+	{
+		.name = "raw",
+		.show = hmc630x_iio_read_raw,
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_channel hmc630x_channels[] = {
+	{
+		.ch_type = IIO_TEMP,
+		.attributes = hmc630x_iio_temp_attrs,
+	}
+};
+
+#define HMC630X_COMMON_ATTRS \
+	{ \
+		.name = "enabled", \
+		.priv = HMC630X_IIO_ATTR_ENABLED, \
+		.show = hmc630x_iio_read_attr, \
+		.store = hmc630x_iio_write_attr, \
+	}, \
+	{ \
+		.name = "vco", \
+		.priv = HMC630X_IIO_ATTR_VCO, \
+		.show = hmc630x_iio_read_attr, \
+		.store = hmc630x_iio_write_attr, \
+	}, \
+	{ \
+		.name = "vco_available", \
+		.priv = HMC630X_IIO_ATTR_VCO_AVAILABLE, \
+		.show = hmc630x_iio_read_vco_available, \
+	}, \
+	{ \
+		.name = "if_attn", \
+		.priv = HMC630X_IIO_ATTR_IF_ATTN, \
+		.show = hmc630x_iio_read_attr, \
+		.store = hmc630x_iio_write_attr, \
+	}, \
+	{ \
+		.name = "temp_en", \
+		.priv = HMC630X_IIO_ATTR_TEMP_EN, \
+		.show = hmc630x_iio_read_attr, \
+		.store = hmc630x_iio_write_attr, \
+	}
+
+static struct iio_attribute hmc6300_iio_attrs[] = {
+	HMC630X_COMMON_ATTRS,
+	{
+		.name = "rf_attn",
+		.priv = HMC630X_IIO_ATTR_RF_ATTN,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute hmc6301_iio_attrs[] = {
+	HMC630X_COMMON_ATTRS,
+	{
+		.name = "rf_lna_gain",
+		.priv = HMC630X_IIO_ATTR_RF_LNA_GAIN,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	{
+		.name = "bb_attn1",
+		.priv = HMC630X_IIO_ATTR_BB_ATTN1,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	{
+		.name = "bb_attn2",
+		.priv = HMC630X_IIO_ATTR_BB_ATTN2,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	{
+		.name = "bb_attni_fine",
+		.priv = HMC630X_IIO_ATTR_BB_ATTNI_FINE,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	{
+		.name = "bb_attnq_fine",
+		.priv = HMC630X_IIO_ATTR_BB_ATTNQ_FINE,
+		.show = hmc630x_iio_read_attr,
+		.store = hmc630x_iio_write_attr,
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_device hmc630x_iio_device_template = {
+	.attributes = NULL, // filled at run-time
+	.num_ch = NO_OS_ARRAY_SIZE(hmc630x_channels),
+	.channels = hmc630x_channels,
+	.debug_reg_read = (int32_t (*)())_hmc630x_read_register2,
+	.debug_reg_write = (int32_t (*)())_hmc630x_write_register2
+};
+
+int32_t hmc630x_iio_init(struct hmc630x_iio_dev **iiodev,
+			 struct hmc630x_iio_init_param *init_param)
+{
+	int ret;
+	struct hmc630x_iio_dev *d;
+	struct iio_device *d2;
+
+	d = (struct hmc630x_iio_dev *)calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	d2 = (struct iio_device *)calloc(1, sizeof(*d2));
+	if (!d2)
+		goto end_0;
+
+	*d2 = hmc630x_iio_device_template;
+
+	d->iio_dev = d2;
```

Comment:
`d->iio_dev = &hmc630x_iio_device_template` and then you don't need to allocate `d2` anymore.

Reply:
And what if you have multiple devices ? You're going the end up with the same d->iio_dev for all of them, which is not ok.